Adding documentation, delete unnecessary fn and argument (#4128)
Fusion-Power-Plant-Framework/bluemira: documentation/source/equilibria/equilibria.rst

Toroidal Harmonic constraints
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
An equilibrium poloidal field has a plasma and coilset contribution. When we optimise coil currents,
we also have to re-solve the equilbrium solution for the plasma. We can use the following premise that
our equilibrium solution will not change if the coilset contribution
to the poloidal field (vacuum field) is kept the same in the region
occupied by the core plasma (i.e. the region characterised by closed flux surfaces).
If we constrain the core plasma while altering (optimising) other aspects
of the magnetic configuration, then we will not need to re-solve for
the plasma equilibrium at each iteration. We can decompose the vacuum field into Toroidal Harmonics (TH)
to create a minimal set of constraints for use in optimisation.

There are benefits to using TH as a minimal set of constraints:
- We can choose not to re-solve for the plasma equilibrium at each step, since the
coilset contribution to the core plasma (within the LCFS) is constrained.
- We have a minimal set of constraints (a set of harmonic amplitudes) for the core
plasma contribution, which can reduce the dimensionality of the problem we are
considering.

> We get the TH amplitudes/coefficients, :math:`A(\tau, \sigma)`, from the following equations:

Reviewer: ```suggestion
An equilibrium poloidal field has a plasma and coilset contribution.We can use Toroidal Harmonic (TH) functions to approximate the coilset contribution to the poloidal magnetic flux using equations :eq:` ToroidalHarmonics `and :eq:` PoloidalFlux `.
```

> 

Reviewer: ```suggestion

.. math::
   :label: PoloidalFlux
   \psi = R A = \frac{R_0 \sinh \tau}{\Delta} A 

```

.. math::

>    :label: TH_no_currents

Reviewer: ```suggestion
   :label: ToroidalHarmonics
```

   \begin{aligned}
   A(\tau, \sigma) &= \sum_{m=0}^{\infty} A_m^{\cos} \epsilon_m m! \sqrt{\frac{2}{\pi}}
   \Delta^{\frac{1}{2}} \textbf{Q}_{m-\frac{1}{2}}^{1}(\cosh \tau) \cos(m \sigma)  \\
   &+ A_m^
   {\sin}
   \epsilon_m m! \sqrt{\frac{2}{\pi}} \Delta^{\frac{1}{2}}
   \textbf{Q}_{m-\frac{1}{2}}^{1}(\cosh \tau) \sin(m \sigma)\end{aligned}

where

> .. math::
>    :label: TH_with_currents
> 
>    A_m^{\cos, \sin} = \frac{\mu_0 I_c}{2^{\frac{5}{2}}} factorial\_term \frac{\sinh(
>    \tau_c)}
>    {\Delta_c^{\frac{1}{2}}} P_{m - \frac{1}{2}}^{-1}(\cosh(\tau_c)) ^{\cos}_{\sin}(m
>    \sigma_c)
> 
> where
> 

Reviewer: ```suggestion
```

- :math:`A_m^{\cos, \sin}` are coefficients for a single coil

- subscript :math:`c` refers to a single coil

- :math:`I_c, \tau_c, \sigma_c` are the coil current, and coil position in toroidal coordinates :math:`(\tau, \sigma)`

- :math:`m` is the poloidal mode number

- :math:`P_{\nu}^{\mu}` is the associated Legendre function of the first kind of degree :math:`\nu` and order :math:`\mu`

- :math:`\textbf{Q}_{\nu}^{\mu}` is Olver's definition of the associated Legendre function of the second kind. See [here](https://dlmf.nist.gov/14) or F. W. J. Olver (1997b) Asymptotics and Special Functions. A. K. Peters, Wellesley, MA. for more information.

- :math:`\varepsilon_m = 1` for :math:`m = 0` and :math:`\varepsilon_m = 2` for :math:`m \ge 1`

- :math:`\Delta = \cosh(\tau) - \cos(\sigma)`

- :math:`\Delta_c = \cosh(\tau_c) - \cos(\sigma_c)`

> - :math:`factorial\_term = \prod_{i=0}^{m-1} \left( 1 + \frac{1}{2(m-i)}\right)`

Reviewer: ```suggestion
- :math:`\psi` is poloidal flux

- :math:`m` is the poloidal mode number

- :math:`A_m^{\cos, \sin}` are harmonic amplitudes

- :math:`\varepsilon_m = 1` for :math:`m = 0` and :math:`\varepsilon_m = 2` for :math:`m \ge 1`

- :math:`\Delta = \cosh(\tau) - \cos(\sigma)`

- :math:`\textbf{Q}_{\nu}^{\mu}` is Olver's definition of the associated Legendre function of the second kind.

Toroidal coordinates are defined as :math:`\tau = \ln\frac{d_1}{d_2}` and :math:`\sigma  = sign(z - z_0) \arccos\frac{d_1^2 + d_2^2 - 4 R_0^2}{2 d_1 d_2}`where, :math:`d_1^2 = (R + R_0)^2 + (z - z_0)^2`and :math: :`d_2^2 = (R - R_0)^2 + (z - z_0)^2`.
```

up the region to be used in the TH approximation using the `toroidal_harmonic_grid_and_coil_setup` function.
This function needs to be provided with the equilbrium, and the coordinates of the focus point. By default the
focus point is set to the effective centre of the plasma. There are also optional arguments which are used
to specify where the approximation region is placed. The `tau_limit` is used to determine how the maximum
tau value is chosen. The three options are:

- LCFS: use the maximum extent of the LCFS
- COIL: use the maxmimum area that is contained within all coils
- MANUAL: use a user-specified tau limit

The `min_tau_value` is used to specify the minimum tau for the toroidal coordinate approximation region, and
lower min tau means a larger region of space (the maximum tau is at the focus point).

Then it iterates through combinations of cos and sin modes,
calculates the psi at the collocation points using equation :eq:`TH_no_currents`, compares this approximated psi to the bluemira psi at the collocation points,
and returns the combination of modes that gives the best approximation when compared using an L2 norm of the error across the

> psi map.

Reviewer: ```suggestion
Toroidal harmonic amplitudes can be implemented as a set of constraints or targets in a coilset optimisation. The TH amplitudes are used to preserve the coilset contribution in the region occupied by the core plasma (i.e. the region characterised by closed flux surfaces), enabling modification of the poloidal magnetic configuration outside the core region without (significantly) altering the plasma equilibrium.

Potential benefits to using TH as a set of constraints include:
- We can choose not to re-solve for the plasma equilibrium at each optimisation step, since the
coilset contribution to the core plasma (within the LCFS) is constrained.
- We have a minimal set of constraints (a set of harmonic amplitudes) for the core
plasma contribution, which can reduce the dimensionality of the problem we are
considering.

To set up a toroidal harmonic constraint, we first find the significant poloidal modes and their amplitude values using equations  :eq:` ToroidalHarmonics `and :eq:` PoloidalFlux `.

The user chooses the: 
- Number of degrees of freedom (DoF) appropriate for the given optimisation problem – this limits the number of poloidal modes that can be used in the TH approximation. The default value will allow DoF up to the number of coilset currents being optimised.  
- Normalised psi value of the flux surface within which the coilset contribution to psi will be constrained.

Then these values are input in `toroidal_harmonic_approximation`, which iterates through all the combinations of cos and sin modes up to the DoF limit. For each iteration:
- Psi values for a set of sampled positions within the selected flux surface are found and used to determine the amplitudes of the contributing poloidal modes.
- L2 norm of the error between the approximated coilset psi and the true coilset psi is calculated. 

The `toroidal_harmonic_approximation` returns the result as a `ToroidalHarmonicsSelectionResult`for the approximate psi with the smallest difference to the true coilset psi in the selected closed flux region. This contains all the information needed to set up a `ToroidalHarmonicConstraint`, as well as supporting information that can be used in analysis.
```

> When we want to use TH as a constraint, we use equation :eq:`TH_with_currents` to create an :math:`A\bf{x} = b` constraint.

Reviewer: ```suggestion
A `ToroidalHarmonicsSelectionResult` contains the combination of modes (and their amplitude values) that gives the best approximation when compared using an L2 norm of the error across the psi map. These can be then implemented as an A\bf{x} = b constraint or magnetic target using the equation :eq:`THAmplitudeCurrentRelation`.
```

> 

Reviewer: ```suggestion

.. math::
   :label: THAmplitudeCurrentRelation
   A_m^{\cos, \sin} = \frac{\mu_0 I_c}{2^{\frac{5}{2}}} factorial\_term \frac{\sinh(
   \tau_c)}
   {\Delta_c^{\frac{1}{2}}} P_{m - \frac{1}{2}}^{-1}(\cosh(\tau_c)) ^{\cos}_{\sin}(m
   \sigma_c)

where

- :math:`A_m^{\cos, \sin}` are hamonic amplitudes for a single coil

- subscript :math:`c` refers to a single coil

- :math:`I_c, \tau_c, \sigma_c` are the coil current, and coil position in toroidal coordinates.

- :math:`P_{m - \frac{1}{2}}^{-1}` is the associated Legendre function of the first kind. 

- :math:`\Delta_c = \cosh(\tau_c) - \cos(\sigma_c)`

- :math:`factorial\_term = \prod_{i=0}^{m-1} \left( 1 + \frac{1}{2(m-i)}\right)`

```

.. figure:: th-flux-comparison.png
   :name: fig:th-flux-comparison

>    Diagram showing the comparison of coilset psi calculated by bluemira vs calculated using toroidal harmonic equations.

Reviewer: ```suggestion
   Diagram showing a comparison of coilset psi calculated in bluemira and the approximation of coilset psi found using toroidal harmonic functions.
```

> 

Reviewer: ```suggestion

See [here](https://dlmf.nist.gov/14) or F. W. J. Olver (1997b) Asymptotics and Special Functions. A. K. Peters, Wellesley, MA. for more information on the Legendre functions used in the TH approximation.

Note 1:

The default region to be conserved in the TH approximation is set to encompass the the maximum extent of the LCFS, with its focus point at the effective centre of the plasma. However, the function `toroidal_harmonic_grid_and_coil_setup` can be used to approximate a region with a different area and focus point. The optional arguments, which are used to specify where the approximation region is placed, are:
- the maximum extent of the LCFS,
- the maxmimum area that is contained within all coils,
- a user-specified tau limit.
Lower min tau means a larger region of space (the maximum tau is at the focus point).

Note 2:

If a coil is within the TH approximation region then its contribution must be kept fixed during an optimisation. The coilset 'control coils' must be set to be the same as the `th_coilset` returned in the `ToroidalHarmonicsSelectionResult` while setting up the optimisation problem. Please see 'examples/equilibria/Toroidal_Harmonics_Optimisation_Set_Up.ex.py' for an example.

```
